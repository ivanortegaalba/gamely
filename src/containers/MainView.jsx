import React, { Component, Fragment } from 'react'
import { connect } from 'react-redux'
import PropTypes from '../PropTypes'
import styled from 'styled-components'
import { Header, px2rems } from '../style/styled-components'
import { palette, key } from '../style/styled-components/tools/functions'
import { dispatchers } from '../actions'
import { selectors } from '../reducers'
import { GamesSection } from '../components'

// TODO: Move this component to the styled-components library
const InputHeader = styled.input`
  border: 1px solid ${palette(0)};
  color: white;
  text-decoration-color: white;
  padding: ${px2rems(15)};
  width: 50%;
  font-size: ${key('fontSizes.s')}px;
  background-color: ${palette(2)};
  text-align: center;
  font-size: ${props => props.theme};
  /* Chrome, Firefox, Opera, Safari 10.1+ */
  ::placeholder { 
    color: ${key('palette.white')};
    opacity: 1; /* Firefox */
}
`

class MainView extends Component {
  componentDidMount () {
    this.props.fetchGames()
  }

  onSearchChange = (event) => {
    this.props.filterGameBy(event.target.value)
  }

  render () {
    const { games, favourites, filterBy, ...others } = this.props
    return <Fragment>
      <section id='search'>
        <Header palette={'dark'}>
          <InputHeader
            palette={'dark'}
            type='text'
            placeholder={'Search game...'}
            onChange={this.onSearchChange} />
        </Header>
      </section>
      {favourites.length > 0 &&
      <GamesSection title='Recently Saved' games={favourites} {...others}
                    filterGamesStartWith={filterBy} />}
      {games.length > 0 &&
      <GamesSection title='Explore' games={games} {...others} filterGamesStartWith={filterBy}
                    defaultOpen />}
    </Fragment>
  }
}

MainView.propTypes = {
  games: PropTypes.arrayOf(PropTypes.game).isRequired,
  fetchGames: PropTypes.func.isRequired,
  addToFavourites: PropTypes.func.isRequired,
  removeFromFavourites: PropTypes.func.isRequired
}

const mapStateToProps = state => ({
  games: selectors.getAllGames(state),
  favourites: selectors.getFavouriteGames(state),
  filterBy: selectors.getFilterBy(state)
})

const { addToFavourites, removeFromFavourites, fetchGames, filterGameBy } = dispatchers
export default connect(
  mapStateToProps,
  {
    addToFavourites,
    removeFromFavourites,
    fetchGames,
    filterGameBy
  }
)(MainView)
