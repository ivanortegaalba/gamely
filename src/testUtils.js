import 'jest-styled-components'
import React from 'react'
import { shallow } from 'enzyme'
import { ThemeProvider } from 'styled-components'



export const shallowWithTheme = (tree, theme) => {
  const context = shallow(<ThemeProvider theme={theme} />)
    .instance()
    .getChildContext()
  return shallow(tree, { context })
}
