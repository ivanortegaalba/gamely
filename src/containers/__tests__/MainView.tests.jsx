import React from 'react'
import { shallowWithStore } from '../../testUtils'
import MainView from '../MainView'
import { createMockStore } from 'redux-test-utils'

// TODO: Test the component without connect
// TODO: Test the componentDidMount lifecycle
describe('<MainView />', () => {
  describe('init', () => {
    const store = createMockStore({
      entities: {},
      favourites: [],
      games: [],
      filterBy: ''
    })
    const component = <MainView />
    const wrapper = shallowWithStore(component, store)
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
    it('should pass connect the necessary props', () => {
      expect(wrapper.props()).toMatchSnapshot()
    })
  })
})
