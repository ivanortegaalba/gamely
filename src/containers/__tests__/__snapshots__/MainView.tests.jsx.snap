// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainView /> init should pass connect the necessary props 1`] = `
Object {
  "addToFavourites": [Function],
  "favourites": Array [],
  "fetchGames": [Function],
  "filterBy": "",
  "filterGameBy": [Function],
  "games": Array [],
  "removeFromFavourites": [Function],
}
`;

exports[`<MainView /> init should render correctly 1`] = `
<MainView
  addToFavourites={[Function]}
  favourites={Array []}
  fetchGames={[Function]}
  filterBy=""
  filterGameBy={[Function]}
  games={Array []}
  removeFromFavourites={[Function]}
/>
`;
